README.md
# Password Storing Daemon (passwordd)
Setting the password is done by sending a json object:
```
{"action":"set", "key":"<key name>", "value":"<secret>"}
```


Retrieving the password is done the same way, but now action is
'get' and no value is sent:
```
{"action":"get", "key":"<key name>"}
```

The response is also a json object containing of two entries:
```
{"result":"<result>", "value":"<value>"}
```
The `<result>` can either be:
 - "ok" : everything went fine
 - "error" : an error occured

The `<value>` contains the secret in case of a successful get request,
else it is just empty.


An example netcat session:
```
$ netcat 127.0.0.1 6969
{"action":"set", "key":"test", "value":"test123"}
{"result":"ok","value":""}

$ netcat 127.0.0.1 6969
{"action":"get", "key":"test"}
{"result":"ok","value":"test123"}
```


debian/changelog
passwordd (2.1.11) UNRELEASED; urgency=medium

  * Initial release.
  * Improve initscripts
  * postrm -> prerm
  * fix install and upgrade terror 
  * next try
  * test
  * Automatically start passwordd
  * fix typo in passworddsync
  * fix further errors in passworddsync, initscripts
  * improve output of passwordc
  * adjust config for passworddsync
  * add systemd script for password service 
  * Integrate fixes from first runs
  * Merge all updates into main branch
  * Move towards systemd
  * Add post starup delay to systemd unit file
  * Extended configurability of passworddsync
  * Add dependence on network-pingable

 -- Marcus Hardt <marcus@hardt-it.de>  Thu 10 Jun 2021 11:44:11 AM CEST


passwordc/passwordc.go
package main

import (
	//"github.com/watts-kit/passwordd/passwordclib"
	"../passwordclib"

	"bufio"
	"gopkg.in/alecthomas/kingpin.v2"
	"fmt"
	"os"
	"strings"
)

const(
	version = "1.0.0"
)


var (
	versionText = "cli:"+version+", lib:"+passwordclib.Version()
	app = kingpin.New("passwordc",
		"a cli client to set/get passwords").Version(versionText)
	set_pwd     = app.Command("set", "set a secret")
	set_pwd_key = set_pwd.Arg("key", "the id of key to set").Required().String()

	get_pwd     = app.Command("get", "get a secret")
	get_pwd_key = get_pwd.Arg("key", "the id of key to get").Required().String()

    overwrite_pwd     = app.Command("overwrite", "replace a secret")
    overwrite_pwd_key = overwrite_pwd.Arg("overwrite", "the id of key to replace").Required().String()
)


func main() {
	switch kingpin.MustParse(app.Parse(os.Args[1:])) {
	case get_pwd.FullCommand():
		password, err := passwordclib.GetPassword(*get_pwd_key)
		if err != nil {
			fmt.Println("error getting password")
			os.Exit(3)
		}
		fmt.Printf("returned secret: '%s'\n", password)
	case set_pwd.FullCommand():
		fmt.Print("please enter the secret: ")
		reader := bufio.NewReader(os.Stdin)
		secret, err := reader.ReadString('\n')
		if err != nil {
			fmt.Println("could not read secret")
			os.Exit(2)
		}
		secret = strings.Trim(secret, "\n")
		_, err = passwordclib.SetPassword(*set_pwd_key, secret)
		if err != nil {
			fmt.Println("could not set secret")
			os.Exit(4)
		}
		fmt.Println("secret set")
	case overwrite_pwd.FullCommand():
		fmt.Print("please enter the secret: ")
		reader := bufio.NewReader(os.Stdin)
		secret, err := reader.ReadString('\n')
		if err != nil {
			fmt.Println("could not read secret")
			os.Exit(2)
		}
		secret = strings.Trim(secret, "\n")
		_, err = passwordclib.OverwritePassword(*overwrite_pwd_key, secret)
		if err != nil {
			fmt.Println("could not replace secret")
            fmt.Println(err)
			os.Exit(4)
		}
		fmt.Println("secret replaced")
	}
}


passwordclib/client.py
#!/usr/bin/env python
import socket
import json


def connect():
    TCP_IP = '127.0.0.1'
    TCP_PORT = 6969
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((TCP_IP, TCP_PORT))
    return s

def receive_and_close(sock):
    BUFFER_SIZE = 1024
    data = sock.recv(BUFFER_SIZE)
    sock.close()
    response = json.loads(data)
    return response

def get_secret(key):
    sock = connect()
    message = json.dumps({"action":"get", "key":key})
    sock.send(message)
    response = receive_and_close(sock)
    if response["result"] == "ok":
        return response["value"]
    return ""

def set_secret(key, secret):
    sock = connect()
    message = json.dumps({"action":"set", "key":key, "value":secret})
    sock.send(message)
    response = receive_and_close(sock)
    if response["result"] == "ok":
        return "secret set"
    return "secret NOT set"

def overwrite_secret(key, secret):
    sock = connect()
    message = json.dumps({"action":"overwrite", "key":key, "value":secret})
    sock.send(message)
    response = receive_and_close(sock)
    if response["result"] == "ok":
        return "secret set"
    return "secret NOT set"



def main():
    print set_secret("python_test", "this python test was good :)")
    print get_secret("python_test")

if __name__ == "__main__":
    main()


passwordclib/passwordclib.go
package passwordclib

import (
	"encoding/json"
	"net"
)

func GetPassword(key string) (string, error) {
	conn := connect()
	return sendAndCheckResponse(Request{"get", key, ""}, conn)
}

func SetPassword(key string, value string) (string, error) {
	conn := connect()
	return sendAndCheckResponse(Request{"set", key, value}, conn)
}

func OverwritePassword(key string, value string) (string, error) {
	conn := connect()
	return sendAndCheckResponse(Request{"overwrite", key, value}, conn)
}

func Version() string {
	return version
}

const (
	port    = "127.0.0.1:6969"
	version = "1.0.0"
)

type Error struct {
	Description string
}

func (e Error) Error() string {
	return e.Description
}

type Request struct {
	Action string `json:"action"`
	Key    string `json:"key"`
	Value  string `json:"value"`
}

type Response struct {
	Result string `json:"result"`
	Value  string `json:"value"`
}


func connect() net.Conn {
	c, err := net.Dial("tcp", port)
	if err != nil {
		return nil
	}
	return c
}

func sendAndCheckResponse(request Request, conn net.Conn) (string, error) {
	if conn == nil {
		return "", Error{"could not connect"}
	}
	defer conn.Close()
	sendRequest(request, conn)
	return readResponse(conn)
}

func sendRequest(request Request, conn net.Conn) {
	data_out, _ := json.Marshal(request)
	conn.Write(data_out)
}

func readResponse(conn net.Conn) (string, error) {
	response := Response{"error", ""}
	buf := make([]byte, 4096)
	numbytes, err := conn.Read(buf)
	if numbytes == 0 {
		return "", Error{"no data received"}
	}
	if err != nil {
		return "", err
	}
	jsonErr := json.Unmarshal(buf[:numbytes], &response)
	if jsonErr != nil {
		return "", jsonErr
	}
	if response.Result != "ok" {
		return "", Error{"received result: "+response.Result}
	}
	return response.Value, nil
}


passworddsync/passworddsync.py
#!/usr/bin/env python
# pylint: disable=invalid-name
import socket
import json
import sys
import os
# apt-get install python-configargparse
import argparse

def parseOptions():
    '''Parse the commandline options'''
    path_of_executable = sys.argv[0]
    folder_of_executable = os.path.split(path_of_executable)[0]

    parser = argparse.ArgumentParser(
        description='''Synchronisation of password daemons''')
    parser.add_argument('--pullhost', '--pull',
                        help='hostname to pull passwords from',
                        default='127.0.0.1')
    parser.add_argument('--pullport', '--port',
                        help='port to pull passwords from',
                        default=6969, type=int)
    parser.add_argument('--pushhost', '--push',
                        help='hostname to push passwords to',
                        default='127.0.0.1')
    parser.add_argument('--verbose', '-v', action='count')
    parser.add_argument('--pushport',
                        help='port to push passwords to',
                        default=6969, type=int)
    parser.add_argument('key', action='append', nargs='+',
                        help='List of keys to operate on')
    return parser.parse_args()

def connect(host_port):
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect(host_port)
    return s

def receive_and_close(sock):
    BUFFER_SIZE = 1024
    data = sock.recv(BUFFER_SIZE)
    sock.close()
    response = json.loads(data)
    return response

def get_secret(host_port, key):
    if args.verbose > 0:
        print ("connecting to: %s" % str(host_port))
    sock = connect(host_port)
    message = json.dumps({"action":"get", "key":key})
    sock.send(message)
    if args.verbose > 2:
        print ("sending message: '%s'" % message)
    response = receive_and_close(sock)
    if args.verbose > 2:
        print ("Response: '%s'" % response)
    if response["result"] == "ok":
        return response["value"]
    return ""

def set_secret(host_port, key, secret):
    sock = connect(host_port)
    message = json.dumps({"action":"set", "key":key, "value":secret})
    sock.send(message)
    response = receive_and_close(sock)
    if response["result"] == "ok":
        return "secret set"
    return "secret NOT set"

def overwrite_secret(host_port, key, secret):
    sock = connect(host_port)
    message = json.dumps({"action":"overwrite", "key":key, "value":secret})
    sock.send(message)
    if args.verbose > 2:
        print ("sending message: '%s'" % message)
    response = receive_and_close(sock)
    if response["result"] == "ok":
        return "secret set"
    return "secret NOT set"


args = parseOptions()
def main():
    if args.verbose > 0:
        print("keys: %s" % str(args.key))

    source = (args.pullhost, args.pullport)
    destin = (args.pushhost, args.pushport)
    if args.verbose > 1:
        print('Source: "%s"'%str(source))
        print('Destin: "%s"'%str(destin))

    for key in args.key[0]:
        # check if key is available in destination
        dst_secret = get_secret(destin, key)
        if args.verbose > 2:
            print ("got dst_secret: %s" % dst_secret)
        # if dst_secret: # i.e. if overwriting is not intended 
        #     print('Secret "%s" is already available in destination at %s:%d' %
        #           (key, args.pullhost, args.pullport))
        #     # -=> skip to next
        #     continue
        if args.verbose > 1:
            print ('    Got dest:%s' % dst_secret)
        # check if key is available in source
        src_secret = get_secret(source, key)
        if args.verbose > 2:
            print ('    Got src: %s' % src_secret)
        if not src_secret:
            print('Cannot get secret for "%s" from %s:%d' %
                  (key, args.pullhost, args.pullport))
            # -=> warning and skip to next
            continue
        if args.verbose > 2:
            print ('    Got src: %s' % src_secret)
        # get in source // happened above
        # set in destination
        result = set_secret(destin, key, src_secret)
        if args.verbose > 0:
            print ("setting new secret. Result: '%s' " % result)
        if result == 'secret NOT set':
            if args.verbose > 0:
                print ("setting didn't work, overwriting")
            result = overwrite_secret(destin, key, src_secret)
            if args.verbose > 0:
                print ("overwriting old secret. Result: '%s' " % result)

        if args.verbose > 0:
            print('copied secret for "%s"' % key)

if __name__ == "__main__":
    main()


passwordd.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/takama/daemon"
	"log"
	"net"
	"os"
	"os/signal"
	"syscall"
)

const (

	// name of the service
	name        = "passwordd"
	description = "Simple Password storing daemon"

	// port which daemon should be listen
	//port = "127.0.0.1:6969"
	port = "0.0.0.0:6969"

	// the current version
	version = "1.0.1"
)

// dependencies that are NOT required by the service, but might be used
var dependencies = []string{}

var stdlog, errlog *log.Logger

// Service has embedded daemon
type Service struct {
	daemon.Daemon
}

type Request struct {
	Action string `json:"action"`
	Key    string `json:"key"`
	Value  string `json:"value"`
}

type Response struct {
	Result string `json:"result"`
	Value  string `json:"value"`
}

// Manage by daemon commands or run the daemon
func (service *Service) Manage() (string, error) {

	usage := "Usage: pwd-daemon install | remove | start | stop | status"

	// if received any kind of command, do it
	// else just run the daemon on the console
	if len(os.Args) > 1 {
		command := os.Args[1]
		switch command {
		case "install":
			return service.Install()
		case "remove":
			return service.Remove()
		case "start":
			return service.Start()
		case "stop":
			return service.Stop()
		case "status":
			return service.Status()
		case "version":
			return version, nil
		default:
			return usage, nil
		}
	}
	// Set up channel on which to send signal notifications.
	// We must use a buffered channel or risk missing the signal
	// if we're not ready to receive when the signal is sent.
	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt, os.Kill, syscall.SIGTERM)

	// Set up listener for defined host and port
	listener, err := net.Listen("tcp", port)
	if err != nil {
		return "Possibly was a problem with the port binding", err
	}

	// set up channel on which to send accepted connections
	listen := make(chan net.Conn, 100)
	go acceptConnection(listener, listen)

	passwords := make(map[string]string)
	// loop work cycle with accept connections or interrupt
	// by system signal
	for {
		select {
		case conn := <-listen:
			go handleClient(conn, passwords)

		case killSignal := <-interrupt:
			stdlog.Println("Got signal:", killSignal)
			stdlog.Println("Stoping listening on ", listener.Addr())
			listener.Close()
			if killSignal == os.Interrupt {
				return "Daemon was interrupted by system signal", nil
			}
			return "Daemon was killed", nil
		}
	}
}

// Accept a client connection and collect it in a channel
func acceptConnection(listener net.Listener, listen chan<- net.Conn) {
	for {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		listen <- conn
	}
}

func handleClient(client net.Conn, passwords map[string]string) {
	for {
		request := &Request{}
		bad_response := &Response{"error", ""}
		buf := make([]byte, 4096)
		numbytes, err := client.Read(buf)
		error_data_out, _ := json.Marshal(bad_response)
		if numbytes == 0 || err != nil {
			client.Write(error_data_out)
			client.Close()
			return
		}

		jsonErr := json.Unmarshal(buf[:numbytes], &request)
		if jsonErr != nil {
			client.Write(error_data_out)
			client.Close()
			return
		}

		if request.Action == "get" {
			secret, exists := passwords[request.Key]
			good_response := &Response{"ok", secret}
			good_data_out, _ := json.Marshal(good_response)
			if exists == true {
				client.Write(good_data_out)
			} else {
				client.Write(error_data_out)
			}
		} else if request.Action == "set" {
			_, exists := passwords[request.Key]
			good_response := &Response{"ok", ""}
			good_data_out, _ := json.Marshal(good_response)
			if exists == true {
				client.Write(error_data_out)
			} else {
				passwords[request.Key] = request.Value
				client.Write(good_data_out)
			}
		} else if request.Action == "overwrite" {
			_, exists := passwords[request.Key]
			good_response := &Response{"ok", ""}
			good_data_out, _ := json.Marshal(good_response)
			if exists == true {
				delete(passwords, request.Key)
				passwords[request.Key] = request.Value
				client.Write(good_data_out)
			}
		} else {

			client.Write(error_data_out)
		}
		client.Close()
		return

	}
}

func init() {
	stdlog = log.New(os.Stdout, "", 0)
	errlog = log.New(os.Stderr, "", 0)
}

func main() {
	srv, err := daemon.New(name, description, daemon.SystemDaemon, dependencies...)
	if err != nil {
		errlog.Println("Error: ", err)
		os.Exit(1)
	}
	service := &Service{srv}
	status, err := service.Manage()
	if err != nil {
		errlog.Println(status, "\nError: ", err)
		os.Exit(1)
	}
	fmt.Println(status)
}
